src/coredump/coredump.c

```diff
@@ -68,28 +68,57 @@
 
 enum {
-        /* We use this as array indexes for a couple of special fields we use for naming
-         * coredump files, and attaching xattrs, and for indexing argv[].
+        /* We use these as array indexes for our process metadata cache.
          *
-         * In the internal context[] array, fields before CONTEXT_COMM are the strings
-         * from argv[] passed by the kernel according to our pattern defined in
-         * /proc/sys/kernel/core_pattern (see man:core(5)). So they should not be
-         * freed. The strings at indices CONTEXT_COMM and higher are allocated by us and
-         * should be freed at the end.
-         */
-        CONTEXT_PID,
-        CONTEXT_UID,
-        CONTEXT_GID,
-        CONTEXT_SIGNAL,
-        CONTEXT_TIMESTAMP,
-        CONTEXT_RLIMIT,
-        CONTEXT_HOSTNAME,
-        CONTEXT_COMM,
-        CONTEXT_EXE,
-        CONTEXT_UNIT,
-        _CONTEXT_MAX
+         * The first indices of the cache stores the same metadata as the ones passed by
+         * the kernel via argv[], ie the strings array passed by the kernel according to
+         * our pattern defined in /proc/sys/kernel/core_pattern (see man:core(5)). */
+
+        META_ARGV_PID,          /* %P: as seen in the initial pid namespace */
+        META_ARGV_UID,          /* %u: as seen in the initial user namespace */
+        META_ARGV_GID,          /* %g: as seen in the initial user namespace */
+        META_ARGV_SIGNAL,       /* %s: number of signal causing dump */
+        META_ARGV_TIMESTAMP,    /* %t: time of dump, expressed as seconds since the Epoch */
+        META_ARGV_RLIMIT,       /* %c: core file size soft resource limit */
+        META_ARGV_HOSTNAME,     /* %h: hostname */
+        _META_ARGV_MAX,
+
+        /* The following indexes are cached for a couple of special fields we use (and
+         * thereby need to be retrieved quickly) for naming coredump files, and attaching
+         * xattrs. Unlike the previous ones they are retrieved from the runtime
+         * environment. */
+
+        META_RT_COMM = _META_ARGV_MAX,
+        _META_MANDATORY_MAX,
+
+        /* The rest are similar to the previous ones except that we won't fail if one of
+         * them is missing. */
+
+        META_RT_EXE = _META_MANDATORY_MAX,
+        META_RT_UNIT,
+        _META_MAX
 };
 
+static const char * const meta_field_names[_META_MAX] = {
+        [META_ARGV_PID]          = "COREDUMP_PID=",
+        [META_ARGV_UID]          = "COREDUMP_UID=",
+        [META_ARGV_GID]          = "COREDUMP_GID=",
+        [META_ARGV_SIGNAL]       = "COREDUMP_SIGNAL=",
+        [META_ARGV_TIMESTAMP]    = "COREDUMP_TIMESTAMP=",
+        [META_ARGV_RLIMIT]       = "COREDUMP_RLIMIT=",
+        [META_ARGV_HOSTNAME]     = "COREDUMP_HOSTNAME=",
+        [META_RT_COMM]           = "COREDUMP_COMM=",
+        [META_RT_EXE]            = "COREDUMP_EXE=",
+        [META_RT_UNIT]           = "COREDUMP_UNIT=",
+};
+
+typedef struct ProcessInfoCache {
```

Comment:
why call this "Cache"? i mean, sure, we only read these things once, and then reuse them, but for me a cache is something with LRU logic, and things like that. I would just call this like the array before: "context".

Reply:
Okay